tests/test_semantic_comment.py

```diff
@@ -0,0 +1,39 @@
+
+
+import re
+
+
+def test_semantic_comments():
+    from nb2workflow.semantics import understand_comment_references
+
+    r = understand_comment_references("http://odahub.io/ontology#StartTimeISOT")
+    assert r['owl_type'] == "http://odahub.io/ontology#StartTimeISOT"
+
+    # allow some extra comment
+    r = understand_comment_references("http://odahub.io/ontology#StartTimeISOT . # and some text")
+    # assert r['owl_type'] == "http://odahub.io/ontology#StartTimeISOT"    
+
+    r = understand_comment_references("oda:StartTimeISOT")
+    assert r['owl_type'] == "http://odahub.io/ontology#StartTimeISOT"
+
+
+    normalize = lambda x: re.sub(r"[ \n]+", " ", x).strip()
+
+    r = understand_comment_references("oda:energyMin, oda:keV")
+    assert r['owl_type'] == "http://odahub.io/ontology#energyMin_keV"
+    assert normalize(r['extra_ttl']) == "@prefix oda: <http://odahub.io/ontology#> . oda:energyMin_keV a oda:energyMin, oda:keV ."
+    
+    r = understand_comment_references("oda:energyMin; oda:lower_limit 3;  oda:upper_limit 30")
+    assert r['owl_type'] == "http://odahub.io/ontology#lower_limit_3integer_upper_limit_30integer_energyMin"
```

Comment:
These weird classes are definitely not in the ontology. How to treat it then in the plugin? 
Is there really a need to make such class? Why not return the most general class (e.g. Energy here)?

Probably I just don't understand the idea

Reply:
The most general class does not have the specificity we want to provide.

We want to combine multiple specific properties in a single type, since parameter has a single type here (we could combine attributes from the type in the plugin, but here it combines with other features better below).

It is rather common in this business to use "formal" class/instance names, e.g. with UUID, in rdf reification. Here the name is somewhat meaningful, but it's just for readability, it only needs to be unique. We could in principle change the base to that of the notebook, not generic odahub, I will do this for clarity.

Then, plugin/dispatcher can [make use](https://github.com/oda-hub/dispatcher-app/issues/444) of the ontology relations (e.g. `type`, maybe sometimes `subClassOf`, maybe it should be used more) to infer characteristics. Parameters can combine characteristics of different classes.

Would should make use of the relations in the ontology as we have it defined there https://odahub.io/ontology/. 
But part of the idea here is to complement the ontology as learned from each workflow notebook. 
Each notebook has it's own set of assumptions-statements-triples, which can be also expressed as an rdf graph in this `extra_rdf` attributes, and used in the plugin.

Then, we can sometimes upgrade the notebook-specific classes-instances to odahub.io ontology records.
This is meant to help workflow developers to contribute to the ontology. Right now it's hard, even through we have an interface.
Some terms probably will never end up as common, but it should not hurt since they can be used along with the triples describing them to infer what they mean in each case. So the plugin will have to just do the same thing in every case.


This process is essentially defining equivalent terminology. As much as possible it is good to use common terminology. Sometimes, for a specific case, it is useful to introduce a term for local use. If the local term becomes common and popular, it can be upgraded to official "terminilogy-language" - the odahub.io ontology.

Reply:
Just to keep in mind, when new terms are being introduced, it is essential to pass along their relations to existing terms. Like it's done here in `extra_ttl`.